File: entities/Multiclass.py
from entities.Choice import *

class Multiclass:
    def __init__(self, classe):
        if "prerequisities" in classe:
            self.prerequisitiesAttribute = [item["ability_score"]["index"] for item in classe["prerequisities"] if "prerequisities" in classe]
            self.prerequisitiesValue = [item["minimum_score"] for item in classe["prerequisities"] if "prerequisities" in classe]
            self.numChoices = 0
        elif "prerequisite_options" in classe:
            self.prerequisitiesAttribute = [item["ability_score"]["index"] for item in classe["prerequisite_options"]["from"]["options"]]
            self.prerequisitiesValue = [item["minimum_score"] for item in classe["prerequisite_options"]["from"]["options"]]
            self.numChoices = classe["prerequisite_options"]["choose"]
        self.proficiencies = [item["index"] for item in classe["proficiencies"]]
        self.proficiencyChoices = []

        if "proficiency_choices" in classe:
            for prof in classe["proficiency_choices"]:
                self.proficiencyChoices.append(Choice(prof["choose"], [item["item"]["index"] for item in prof["from"]["options"]]))

File: TrainDataGen/Entities/Multiclass.cs
namespace TrainDataGen.Entities;

public class Multiclass
{
    public List<string> PrerequisitiesAttribute { get; set; }
    public List<int> PrerequisitiesValue { get; set; }
    public int NumChoices { get; set; }
    public List<string> Proficiencies { get; set; }
    public List<Choice> ProficiencyChoices { get; set; }

    public Multiclass(Dictionary<string, object> classe)
    {
        PrerequisitiesAttribute = new List<string>();
        PrerequisitiesValue = new List<int>();
        Proficiencies = new List<string>();
        ProficiencyChoices = new List<Choice>();

        if (classe.ContainsKey("prerequisities"))
        {
            var prereqs = (List<object>)classe["prerequisities"];
            foreach (var itemObj in prereqs)
            {
                var item = (Dictionary<string, object>)itemObj;
                PrerequisitiesAttribute.Add(((Dictionary<string, object>)item["ability_score"])["index"].ToString());
                PrerequisitiesValue.Add((int)item["minimum_score"]);
            }
            NumChoices = 0;
        }
        else if (classe.ContainsKey("prerequisite_options"))
        {
            var options = (Dictionary<string, object>)classe["prerequisite_options"];
            var fromOptions = (List<object>)((Dictionary<string, object>)options["from"])["options"];
            foreach (var itemObj in fromOptions)
            {
                var item = (Dictionary<string, object>)itemObj;
                PrerequisitiesAttribute.Add(((Dictionary<string, object>)item["ability_score"])["index"].ToString());
                PrerequisitiesValue.Add((int)item["minimum_score"]);
            }
            NumChoices = (int)options["choose"];
        }

        if (classe.ContainsKey("proficiencies"))
        {
            var profs = (List<object>)classe["proficiencies"];
            foreach (var itemObj in profs)
            {
                var item = (Dictionary<string, object>)itemObj;
                Proficiencies.Add(item["index"].ToString());
            }
        }

        if (classe.ContainsKey("proficiency_choices"))
        {
            var profChoices = (List<object>)classe["proficiency_choices"];
            foreach (var profObj in profChoices)
            {
                var prof = (Dictionary<string, object>)profObj;
                int choose = (int)prof["choose"];
                var options = (List<object>)((Dictionary<string, object>)prof["from"])["options"];
                var indices = new List<string>();
                foreach (var optionObj in options)
                {
                    var option = (Dictionary<string, object>)optionObj;
                    indices.Add(((Dictionary<string, object>)option["item"])["index"].ToString());
                }
                ProficiencyChoices.Add(new Choice(choose, indices));
            }
        }
    }
}